[SPARK-4859][Core][Streaming] Refactor LiveListenerBus and StreamingListenerBus

zsxwing opened the pull request:
This PR refactors LiveListenerBus and StreamingListenerBus and extracts the common codes to a parent class `ListenerBus`. 

It also includes bug fixes in #3710:
1. Fix the race condition of queueFullErrorMessageLogged in LiveListenerBus and StreamingListenerBus to avoid outputing `queue-full-error` logs multiple times.
2. Make sure the SHUTDOWN message will be delivered to listenerThread, so that we can make sure listenerThread will always be able to exit.
3. Log the error from listener rather than crashing listenerThread in StreamingListenerBus.

During fixing the above bugs, we find it's better to make LiveListenerBus and StreamingListenerBus have the same bahaviors. Then there will be many duplicated codes in LiveListenerBus and StreamingListenerBus. 

Therefore, I extracted their common codes to `ListenerBus` as a parent class: LiveListenerBus and StreamingListenerBus only need to extend `ListenerBus` and implement `onPostEvent` (how to process an event) and `onDropEvent` (do something when droppping an event).

zsxwing committed:
Refactor LiveListenerBus and StreamingListenerBus and make them share same code base

SparkQA commented:
[Test build #25412 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25412/consoleFull) for   PR 4006 at commit [`7cc04c3`](https://github.com/apache/spark/commit/7cc04c385945b490369f9ab0b5831c68ee45a0dc).
 * This patch merges cleanly.

zsxwing commented:
cc @JoshRosen @tdas @andrewor14

zsxwing commented:
Should I open another JIRA, or just update SPARK-4859 jira?

SparkQA commented:
[Test build #25412 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25412/consoleFull) for   PR 4006 at commit [`7cc04c3`](https://github.com/apache/spark/commit/7cc04c385945b490369f9ab0b5831c68ee45a0dc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25412/
Test FAILed.

zsxwing committed:
Remove SparkListenerShutdown

SparkQA commented:
[Test build #25418 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25418/consoleFull) for   PR 4006 at commit [`a9dccd3`](https://github.com/apache/spark/commit/a9dccd372328fd7de739ab2529935563b7c2165d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25418 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25418/consoleFull) for   PR 4006 at commit [`a9dccd3`](https://github.com/apache/spark/commit/a9dccd372328fd7de739ab2529935563b7c2165d).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class ListenerBus[L <: AnyRef, E](name: String) extends ListenerHelper[L, E] `
  * `trait ListenerHelper[L <: AnyRef, E] extends Logging `

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25418/
Test FAILed.

andrewor14 commented:
Hi @zsxwing the same JIRA is fine, but could you add more detail in the issue itself and in this PR description on each of the bugs you are trying to fix? Also, please add a quick motivation of why refactoring is needed in the same places. Thanks.

zsxwing commented:
Updated the desription and jira

zsxwing commented:
Do I need to make some change to passs MiMa tests?

andrewor14 commented:
Yeah, if you look at the test output it'll tell you the exact lines you'll need to add to `MimaExcludes.scala`

zsxwing committed:
Remove final to pass MiMa tests

SparkQA commented:
[Test build #25508 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25508/consoleFull) for   PR 4006 at commit [`4539f91`](https://github.com/apache/spark/commit/4539f918c8da05f9f5f59783441d93d8c10f1d4f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25508 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25508/consoleFull) for   PR 4006 at commit [`4539f91`](https://github.com/apache/spark/commit/4539f918c8da05f9f5f59783441d93d8c10f1d4f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class ListenerBus[L <: AnyRef, E](name: String) extends ListenerHelper[L, E] `
  * `trait ListenerHelper[L <: AnyRef, E] extends Logging `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25508/
Test PASSed.

zsxwing committed:
Add private[spark]

SparkQA commented:
[Test build #25597 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25597/consoleFull) for   PR 4006 at commit [`d2ef990`](https://github.com/apache/spark/commit/d2ef99026095b67eb672a062c1e13370a92cafff).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25597 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25597/consoleFull) for   PR 4006 at commit [`d2ef990`](https://github.com/apache/spark/commit/d2ef99026095b67eb672a062c1e13370a92cafff).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25597/
Test PASSed.

zsxwing committed:
Merge branch 'master' into SPARK-4859-refactor

Conflicts:
	core/src/main/scala/org/apache/spark/scheduler/SparkListenerBus.scala

SparkQA commented:
[Test build #25631 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25631/consoleFull) for   PR 4006 at commit [`0710364`](https://github.com/apache/spark/commit/0710364818d9c1338188d89fa522316d84482ec4).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25631/consoleFull) for   PR 4006 at commit [`0710364`](https://github.com/apache/spark/commit/0710364818d9c1338188d89fa522316d84482ec4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25631/
Test FAILed.

SparkQA commented:
[Test build #25646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25646/consoleFull) for   PR 4006 at commit [`4e85ffc`](https://github.com/apache/spark/commit/4e85ffcec4a8ceb1c379eb92e4932010f405e5c9).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25646/consoleFull) for   PR 4006 at commit [`4e85ffc`](https://github.com/apache/spark/commit/4e85ffcec4a8ceb1c379eb92e4932010f405e5c9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25646/
Test PASSed.

JoshRosen commented:
@tdas @andrewor14 is this PR good to go, or did you have further review comments?

andrewor14 commented:
I still need to do a close review. I will do so within the next day or two.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala:
```diff
@@ -29,113 +28,18 @@ import org.apache.spark.util.Utils
+      // Only log the following message once to avoid duplicated annoying logs.
+      logError("Dropping SparkListenerEvent because no remaining room in event queue. " +
+        "This likely means one of the SparkListeners is too slow and cannot keep up with" +
+        "the rate at which tasks are being started by the scheduler.")
```
you'll need a space after `with` on L40

andrewor14 commented on core/src/main/scala/org/apache/spark/util/ListenerBus.scala:
```diff
@@ -0,0 +1,173 @@
+   * elapsed before the queue emptied.
+   */
+  @VisibleForTesting
+  def waitUntilEmpty(timeoutMillis: Int): Boolean = {
```
do you need the `VisibleForTesting` annotation here since this is already `private[spark]`?

zsxwing replied:
It's just my personal habit. I prefer to use `VisibleForTesting` explicitly rather than documenting it. I saw @aarondav used it in network-shuffle module. So I added `VisibleForTesting` for these methods to make them more explicitly.

andrewor14 commented on core/src/main/scala/org/apache/spark/util/ListenerBus.scala:
```diff
@@ -0,0 +1,173 @@
+   */
+  def stop() {
+    if (!started.get()) {
+      throw new IllegalStateException("Attempted to stop " + name + " that has not yet started!")
```
can you use string interpolation here and other places `s"Attempted to stop $name that has not yet started!"` just to be consistent?

andrewor14 commented:
@zsxwing rather than making a `ListenerHelper` take in two type parameters, can we do the following two things: (1) make `StreamingListenerEvent` extend `SparkListenerEvent`, and (2) make a new trait `Listener` or `ListensToSparkEvents`  and have both `SparkListener` and `StreamingListener` extend it. Would that simplify things or not?

zsxwing committed:
Fix the code style

zsxwing commented:
> (1) make StreamingListenerEvent extend SparkListenerEvent

I don't like it because if so, we will get many meaningless "match may not be exhaustive." warnings from the compiler when we use pattern matching.

My original idea was that creating a parent trait for StreamingListenerEvent and SparkListenerEvent and also a parent trait for SparkListener and StreamingListener. However, it still needs `ListenerHelper` , an event process template, to remove duplicate codes. 

When I was writing `ListenerHelper`, I found I need the type parameters to check the type at the compile time. If not using type parameters, E.g., assume `Listener` is the parent trait of SparkListener and StreamingListener, then the method signature of `addListener` will be `def addListener(listener: Listener)`. The subclasses cannot change it to accept a more specific `Listener`, so they need to do the type check at the runtime to reject a wrong Listener.

Therefore, I added the type parameters and found I could remove the parent traits of StreamingListenerEvent/SparkListenerEvent and SparkListener/StreamingListener. I prefer less classes, so I removed them.

SparkQA commented:
[Test build #26079 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26079/consoleFull) for   PR 4006 at commit [`f0ef647`](https://github.com/apache/spark/commit/f0ef6477e4f1314a0335c3d1ac6a83ca1c4b3d97).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26079 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26079/consoleFull) for   PR 4006 at commit [`f0ef647`](https://github.com/apache/spark/commit/f0ef6477e4f1314a0335c3d1ac6a83ca1c4b3d97).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26079/
Test PASSed.

andrewor14 commented:
I see. That makes sense. However I still don't see the need for a separate `ListenerHelper` trait. In my mind `SparkListenerBus` should extend `ListenerBus[L, E]` instead of `ListenerHelper` directly. If that's the case, then the only class that's actually implementing `ListenerHelper` will be `ListenerBus`, at which point you can just merge the trait with the class itself. Is there a reason for keeping around the trait?

zsxwing commented:
It's because `ReplayListenerBus` needs the functions in SparkListenerBus and ListenerHelper. But it doesn't need `ListenerBus`. That's why `SparkListenerBus` doesn't extend `ListenerBus`.

andrewor14 commented:
@zsxwing I looked at this in greater detail and I have a better understanding of the design choices you made here. The bottom line is that the common properties between `LiveListenerBus` and `StreamingListenerBus` include (1) they both post events in a separate thread asynchronously, and (2) they're both listener buses, but use different types of listeners and events. Therefore, your hierarchy is as follows:
```
                                     ListenerHelper[L, E]
                                      /                          \
                       SparkListenerBus      ListenerBus[L, E]
                          /                \                 /                 \
 ReplayListenerBus         LiveListenerBus      StreamingListenerBus
```
Although this placement of functionality makes sense, I find the naming of the classes very confusing. For instance, it's very strange that `LiveListenerBus` extends both `ListenerBus` and `SparkListenerBus`. From the name, it would appear that `SparkListenerBus` is also an instance of `ListenerBus`, but it is not. Similarly, I would imagine (from the name of the classes) that `ReplayListenerBus` also extends `ListenerBus`, but this is not the case.

So my proposal is the following. Rename your `ListenerHelper` to `ListenerBus`, and your `ListenerBus` to something else, maybe `AsynchronousListenerBus` or something because it posts events to listeners asynchronously in a separate thread.

Do you agree?

zsxwing committed:
Rename ListenerHelper to ListenerBus and the original ListenerBus to AsynchronousListenerBus

SparkQA commented:
[Test build #26279 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26279/consoleFull) for   PR 4006 at commit [`5715061`](https://github.com/apache/spark/commit/571506169c32e9d3d597fa6d2609e7188c5f3748).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26279 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26279/consoleFull) for   PR 4006 at commit [`5715061`](https://github.com/apache/spark/commit/571506169c32e9d3d597fa6d2609e7188c5f3748).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26279/
Test FAILed.

zsxwing commented:
@andrewor14 The news names are pretty good to me. I have renamed them.

zsxwing committed:
Fix the code style after renaming

SparkQA commented:
[Test build #26280 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26280/consoleFull) for   PR 4006 at commit [`c8dade2`](https://github.com/apache/spark/commit/c8dade2aa7c07d97d9dc953e6ea7c09c72ed2f9d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26280 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26280/consoleFull) for   PR 4006 at commit [`c8dade2`](https://github.com/apache/spark/commit/c8dade2aa7c07d97d9dc953e6ea7c09c72ed2f9d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26280/
Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/SparkListenerBus.scala:
```diff
@@ -17,78 +17,47 @@
-   * This does nothing if the event is SparkListenerShutdown.
-   */
-  def postToAll(event: SparkListenerEvent) {
+  def onPostEvent(listener: SparkListener, event: SparkListenerEvent): Unit = {
```
you need override here

andrewor14 commented:
Ok, LGTM I will merge this and fix the override thing myself. Thanks @zsxwing!
